packages/labs/testing/src/lib/fixtures/ssr-fixture.ts

```diff
@@ -0,0 +1,145 @@
+// option for browsers that support declarative shadow DOM as proposed in
+// https://github.com/mfreed7/declarative-shadow-dom/blob/master/README.md#mitigation.
+declare global {
+  interface DOMParser {
+    parseFromString(
+      string: string,
+      type: DOMParserSupportedType,
+      option: {includeShadowRoots: boolean}
+    ): Document;
+  }
+}
+
+/**
+ * Renders the provided Lit template server-side by executing a custom command
+ * for Web Test Runner provided by the Lit SSR Plugin, loads it to the document
+ * and (optionally) hydrates it, returning the element.
+ *
+ * This module **must** be imported before any custom element definitions.
+ */
+export async function ssrFixture<T extends HTMLElement>(
+  template: TemplateResult,
+  {modules, base, hydrate = true}: SsrFixtureOptions
+): Promise<T> {
+  if (base === undefined) {
+    // Find the test file url from the call stack
+    // Chrome:
+    // Error
+    // at ssrFixture (http://localhost:8000/node_modules/@lit-labs/testing/lib/fixtures/ssrFixture.js:29:17)
+    // at ssrHydratedFixture (http://localhost:8000/node_modules/@lit-labs/testing/lib/fixtures/ssrFixture.js:76:12)
+    // at o.<anonymous> (http://localhost:8000/test/my-element_test.js?wtr-session-id=GhB4vW1TXWxXwqgt3F8QD:65:30)
+    //
+    // Firefox:
+    // ssrFixture@http://localhost:8000/node_modules/@lit-labs/testing/lib/fixtures/ssrFixture.js:29:17
+    // ssrHydratedFixture@http://localhost:8000/node_modules/@lit-labs/testing/lib/fixtures/ssrFixture.js:76:12
+    // @http://localhost:8000/test/my-element_test.js?wtr-session-id=NaCljZAFiyeOoV6-qBqwr:65:30
+    //
+    // Webkit:
+    // @http://localhost:8000/node_modules/@lit-labs/testing/lib/fixtures/ssrFixture.js:29:26
+    // asyncFunctionResume@[native code]
+    // asyncFunctionResume@[native code]
+    // @http://localhost:8000/test/my-element_test.js?wtr-session-id=aKWON-wBOBGyzb2CwIvmK:65:37
+    const match = new Error().stack?.match(
+      /http:\/\/localhost.+(?=\?wtr-session-id)/
+    );
+    if (!match) {
+      throw new Error('Could not find call site for ssrFixture');
+    }
+    base = match[0];
+  }
+
+  const rendered = await executeServerCommand<string, Payload>(
+    litSsrPluginCommand,
+    {
+      template,
+      modules: modules.map((module) => new URL(module, base).pathname),
+    }
+  );
+
+  // TODO(augustinekim) Clean up the container from the document
+  const container = document.createElement('div');
+  document.body.appendChild(container);
+
+  if (HTMLTemplateElement.prototype.hasOwnProperty('shadowRoot')) {
+    // Browser natively supports declarative shadowroot.
+    // Shadowroots are only parsed and attached during initial HTML parsing.
+    // innerHTML will not work and must use DOMParser.
+    // See https://web.dev/declarative-shadow-dom/#parser-only
+    const fragment = new DOMParser().parseFromString(
+      // DOMParser removes the opening lit-part comment node
```

Comment:
When I do this:

```ts
new DOMParser().parseFromString("<!--before--><p>hello</p><!--after-->", "text/html")
```

The structure I see is:

```html
<!--before-->
<html>
  <head></head>
  <body>
    <p>hello</p>
    <!--after-->
  </body>
</html>
```

So I think it's not exactly that the comment is ignored, it's that the parser is trying to create a well formed HTML document. It probably keeps the leading comment at the top, because it's allowed to be there, but moves the `<p>` into the `<body>`, because that's the expected well-formed location for it. (The specific element actually matters too -- e.g. a `<p>` will get automatically placed in the `<body>`, while a `<meta>` will get placed in the `<head>`).

So I think your solution is fine, but we could consider doing `<body>${rendered}</body>` and then `fragment.body.childNodes`, to avoid the `querySelector`, and be slightly simpler.